x/dex/keeper/msgserver/msg_server_register_contract.go

```diff
@@ -166,9 +172,43 @@ func (k msgServer) UpdateNewDependencies(ctx sdk.Context, msg *types.MsgRegister
 	return nil
 }
 
-func (k msgServer) SetNewContract(ctx sdk.Context, msg *types.MsgRegisterContract) ([]types.ContractInfo, error) {
+func (k msgServer) HandleDepositOrRefund(ctx sdk.Context, msg *types.MsgRegisterContract) error {
+	creatorAddr, err := sdk.AccAddressFromBech32(msg.Creator)
+	if err != nil {
+		return err
+	}
+	if existingContract, err := k.GetContract(ctx, msg.Contract.ContractAddr); err != nil {
+		// brand new contract
+		if msg.Contract.RentBalance > 0 {
+			if err := k.BankKeeper.SendCoins(ctx, creatorAddr, k.AccountKeeper.GetModuleAddress(types.ModuleName), sdk.NewCoins(sdk.NewCoin(appparams.BaseCoinUnit, sdk.NewInt(int64(msg.Contract.RentBalance))))); err != nil {
+				return err
+			}
+		}
+	} else {
+		if msg.Creator != existingContract.Creator {
+			return sdkerrors.ErrUnauthorized
+		}
+		if msg.Contract.RentBalance < existingContract.RentBalance {
+			// refund
+			refundAmount := existingContract.RentBalance - msg.Contract.RentBalance
+			if err := k.BankKeeper.SendCoins(ctx, k.AccountKeeper.GetModuleAddress(types.ModuleName), creatorAddr, sdk.NewCoins(sdk.NewCoin(appparams.BaseCoinUnit, sdk.NewInt(int64(refundAmount))))); err != nil {
+				return err
+			}
+		} else if msg.Contract.RentBalance > existingContract.RentBalance {
+			// deposit
+			depositAmount := msg.Contract.RentBalance - existingContract.RentBalance
+			if err := k.BankKeeper.SendCoins(ctx, creatorAddr, k.AccountKeeper.GetModuleAddress(types.ModuleName), sdk.NewCoins(sdk.NewCoin(appparams.BaseCoinUnit, sdk.NewInt(int64(depositAmount))))); err != nil {
```

Comment:
nit: we probably want to make this clear in the doc that the creator's account will be charged if they specify a larger than existing balance rentBalance

Reply:
yeah good point